### ISSUE TITLE
Converging in single cluster in DEC (Deep embedded clustering), R implementation

### ISSUE THREAD

**Issue by remahini** [ANTsX/ANTsRNet#27]
Speaking briefly, I noticed a strange performance difference in equal implementations of Deep embedded clustering (DEC) in R and Python (which is available at:  https://github.com/XifengGuo/DEC-keras/blob/master/DEC.py ).

**My question,**
According to the figures below, the **implementation in R** coverage in the stop condition by putting all the observations(i.e. data samples) in one cluster (Figure2).

Here are some evidences:

![Figure1](https://user-images.githubusercontent.com/46174354/77718910-b14b4600-7001-11ea-9524-b7ca0abbf89a.png)
_Figure1 after first initializing with kmeans_
![Figure2](https://user-images.githubusercontent.com/46174354/77718931-bf00cb80-7001-11ea-879b-dafbd1a65ba1.png)
_Figure2 converging in one cluster after the stop condition_
`Train on 70000 samples
Epoch 1/10
70000/70000 [==============================] - 4s 57us/sample - loss: 0.0770
Epoch 2/10
70000/70000 [==============================] - 3s 44us/sample - loss: 0.0444
Epoch 3/10
70000/70000 [==============================] - 3s 44us/sample - loss: 0.0360
Epoch 4/10
70000/70000 [==============================] - 3s 45us/sample - loss: 0.0312
Epoch 5/10
70000/70000 [==============================] - 3s 46us/sample - loss: 0.0279
Epoch 6/10
70000/70000 [==============================] - 3s 44us/sample - loss: 0.0263
Epoch 7/10
70000/70000 [==============================] - 3s 45us/sample - loss: 0.0252
Epoch 8/10
70000/70000 [==============================] - 3s 45us/sample - loss: 0.0243
Epoch 9/10
70000/70000 [==============================] - 3s 45us/sample - loss: 0.0236
Epoch 10/10
70000/70000 [==============================] - 3s 45us/sample - loss: 0.0231
<tensorflow.python.keras.callbacks.History>`

_Figure 3 training AE with 100 epochs_

`ACC =1,  NMI= 1
Iteration 1: (out of 20000): loss = 1e+08, deltaLabel = 0
ACC =0.5117571,  NMI= 0.4666276
Iteration 11: (out of 20000): loss = 0.8442134, deltaLabel = 0.4882429
ACC =0.4964714,  NMI= 0.4784229
Iteration 21: (out of 20000): loss = 0.4245124, deltaLabel = 0.5035286
ACC =0.6418429,  NMI= 0.3421767
Iteration 31: (out of 20000): loss = 1.214075, deltaLabel = 0.3581571
ACC =0.9478857,  NMI= 0.7269456
Iteration 41: (out of 20000): loss = 1.183661, deltaLabel = 0.05211429
ACC =0.6602286,  NMI= 0.8314889
Iteration 51: (out of 20000): loss = 1.712941, deltaLabel = 0.3397714
ACC =0.6393429,  NMI= 0
Iteration 61: (out of 20000): loss = 2.514483, deltaLabel = 0.3606571
ACC =0,  NMI= NaN
Iteration 71: (out of 20000): loss = 2.386619, deltaLabel = 1
ACC =0.9999857,  NMI= 0
Iteration 81: (out of 20000): loss = 2.373431, deltaLabel = 1.428571e-05`

_Figure 4 fitting the DEC model_

Which it was accordingly respect to Figure 5 and Figure 6 in **Python**

![image](https://user-images.githubusercontent.com/46174354/77719894-0be5a180-7004-11ea-91ba-02e24b72757b.png)

_Figure 5 Initializing the labels by kmeans_
[](url)
![image](https://user-images.githubusercontent.com/46174354/77719916-1869fa00-7004-11ea-8aaf-225f22f22863.png)

_Figure 6 final labeling after the model coverage_ (x-axis is the samples  and y-axis is the labels)

Could you please let me know why this is happening? Is **“klb”** really works (What I guess) in this code (R implementation)?

_Information about the machine:
1- Python 3.7 lunched by Spyder 4.1 
2- Rstodio Version 1.2.5033_

The implementation in R : 
https://www.dropbox.com/s/cgavozr1c1rm60s/DEC_test_original.R?dl=0
The implementation in Python:
https://www.dropbox.com/s/c8zd47mv3nsfo3w/DEC_original.py?dl=0

I am waiting for your reply and solution,
Please let me know if any information I can provide!
Thanks for reading!

Sincerely,
Reza

**Comment by ntustison** [comment]
This is certainly the forum for answering specific questions about the code in this repository but it's not a forum for answering general questions about code even if it is related.  In other words, I don't have time to debug other people's code.  I would advise you ask on a more suitable forum such as stack overflow.

**Comment by remahini** [comment]
@ntustison 

Dear Nick Tustison,

Thanks for replying!

Please Let me clear my question which can make a better understanding!

Firstly the code works without any bug. The question is, if the same code and data (i.e. "minst" data in both R and Python codes for DEC) are used, why the DEC R implementation coverage in **one single cluster**? (what cause this?)

Could you please let me know which function in the code makes this problem. 
Cheers!

**Comment by ntustison** [comment]
I'm still debugging *my* code.  You can see that I've made changes yesterday and today.  And I'm still working on my python version in ANTsPyNet.  

When you write "My question is, I have used the same code and data (you included in the code in the comment section), however, the DEC R implementation coverage in one single cluster? (what cause this, this is the question)"    Perhaps it's the English translation but I don't understand this.  It's not a question as it is written.  Maybe try rewording.

**Comment by remahini** [comment]
Dear Nick Tustison,

I am sorry for unclear question. I rephrased my question!
Thanks for devoting time and considering!

I read the changes and debugging in your code. unfotunately, the issue exists while using the dense layers for clustering data (for example the  "mnist" dataset). 

_I got really confused. it should work, but it converges on one single cluster._

could you please let me know what can cause this issue?  I am sorry to take your time but it is very important for me since I have to use it for other data as well.

Kind regards,
Reza

**Comment by remahini** [comment]
Dear Nick Tustison,

Sorry for disturbing you! 

I just want to know what causes the converging of the DEC to a **single cluster**? 
I do not change the parameters you adjusted in your code and just keep them as simple as here:

The implementation in R: https://www.dropbox.com/s/yviyjt4q7bu5wgq/DEC_test_original.R?dl=0
The implementation in Python: https://www.dropbox.com/s/ii3k7rklz7z6446/DEC_original.py?dl=0

I still in **trouble** with DEC when it puts all the samples in **one cluster**.

Could you please let me know what can cause this issue?

I am looking forward to hearing from you,
Thanks for your time!

Sincerely,
Reza

**Comment by ntustison** [comment]
Hi @ReyCJ ,

I don't know why it's converging to a single cluster.  I'm still looking at it.  

Nick

**Comment by remahini** [comment]
Hi Nick!

Thanks for replying!
Hopefully, you will find what is the issue. I am working on too.

Sincerely,
Reza
